tidy3d/components/source.py

```diff
@@ -813,6 +807,72 @@ class AstigmaticGaussianBeam(AngledFieldSource, PlanarSource, BroadbandSource):
 class TFSF(AngledFieldSource, VolumeSource):
     """Total field scattered field with a plane wave field in a volume."""
 
+    injection_axis: Axis = pydantic.Field(
+        ...,
+        title="Injection Axis",
+        description="Specifies the injection axis. The plane of incidence is defined via this "
+        "``injection_axis`` and the ``direction``. The popagation axis is defined with respect "
+        "to the ``injection_axis`` by ``angle_theta`` and ``angle_phi``.",
+    )
+
+    @cached_property
+    def injection_plane_center(self) -> Coordinate:
+        """Center of the injection plane."""
+        sign = 1 if self.direction == "-" else -1
+        center = list(self.center)
+        size = [0 if val == inf else val for val in self.size]
+        center[self.injection_axis] += sign * size[self.injection_axis] / 2
+        return tuple(center)
+
+    def plot(  #  pylint:disable=too-many-arguments
```

Comment:
this looks really similar to `Source.plot` [(code)](https://github.com/flexcompute/tidy3d/blob/2b43faa7c6fa0dd5819b6d962b6835cd56010401/tidy3d/components/source.py#L312-L357). Think it's possible to reuse it? Because it might be annoying to maintain both if we want to change it later.

Reply:
the differences:

## 1. Plotting the geometry
```py
ax = super().plot(x=x, y=y, z=z, ax=ax, **patch_kwargs) # Source
ax = Box.plot(self, x=x, y=y, z=z, ax=ax, **patch_kwargs) # TFSF
```
### proposed solution:
maybe we can just call `Box.plot()` in `Source.plot()` instead of using `super()` and then they should be identical. Actually the Source.plot seems like a not great use of `super()` now that I look at it.

## 2. Custom Center
in `ax = self._plot_arrow()`, `custom_center=self.injection_plane_center` is passed only in `TFSF`

### proposed solutions:
* check if `hasattr(self, "custom_center") in `Source.plot` and add it to `plot_arrow`. (Not ideal)
* add `custom_center` as a `Source` field and just move the whole `TFSF.plot()` to `Source.plot()`
* add `custom_center` as a kwarg in `Source.plot()`,, have `TFSF.plot()` call `Source.plot()` passing `self` as first argument and `self.custom_center` as a kwarg.

Open to other suggestions too, these are just what came to me off the top of my head.

Reply:
On a related note, what are your thoughts on the way the plotting is done right now? The arrow's base is centered on the injection plane rather than the center of the box, which I think makes sense. But in the middle plot, when we're looking at the box "from above" at `z=0`, technically we shouldn't be able to see the arrow because it is located at `z = z_box_bottom`. But at the same time, it's useful to still be able to see the arrow at `z=0`.

![image](https://user-images.githubusercontent.com/97917222/215172087-caf1a556-ae6f-4a8c-b01a-1693fc5a0f1c.png)

Reply:
For consistency, I changed the behaviour so that the arrow is visible only if the plotting plane actually goes through `arrow_base`, so the arrow would not be visible in the center plot, to avoid any confusion.